Add ignoreError feature to dag task

schrodit opened the pull request:
This PR adds the optional attribute "ignoreError" to Argo's DAG task.
If this attribute is set to true the DAG tasks will proceed even if the tagged task fails.

This feature is useful if you want to execute all test and get a result from all of them without the need to depend on successful dependent steps. 
This is necessary as some tests need to be run in parallel but we even want the result of serial task that run afterwards.

As the additional attribute is optional it does not break current workflows.

Here is an example dag workflow where the task B has "continueOn: failed: true" :
![screen shot 2019-02-01 at 14 22 06](https://user-images.githubusercontent.com/7979201/52126252-1fc78a80-262f-11e9-997b-a0a7ac1a8c71.png)

Another one for steps:
<img width="1071" alt="screen shot 2019-02-12 at 15 25 48" src="https://user-images.githubusercontent.com/7979201/52643337-1535ac80-2edd-11e9-901d-33d191068776.png">

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/argoproj/argo?pullRequest=1205) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/argoproj/argo?pullRequest=1205) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/argoproj/argo?pullRequest=1205) it.</sub>

jessesuen commented:
@schrodit thanks for this. it has been a commonly requested feature. @edlee2121 how do you feel about the DSL syntax? One issue I see is that we actually already use the term error and failure separately. And they are meant to have different meanings (failures are under user control, errors happen during initializing/saving). 

I think the we syntax should be clear which case it is handling, or if it is handling both. When thinking about this before, I had thought of the term `bestEffort: true`, but that syntax would also not distinguish errors vs. failures.

We may want to leave this open such as:
```yaml
ignore:
  errors: true
  failures: true
```


Second thing is that we will want to additionally support this in steps, and not just DAGs, because these levels of control are supposed to have feature parity between steps/dags.

edlee2121 commented:
Yes, it would be great if this works for steps as well.
If we already distinguish between failures and errors then I think we should preserve the distinction.

What does everyone think of `ignoreExitStatus`
Trying to find the right balance between precision and conciseness :-)

schrodit commented:
I totally agree that it would make sense to distinguish between errors and failures.
We also have this use case and I think it would make a lot of sense to give the freedom of choice. 

What do you think about about the term `ContinueOn`?
``` 
continueOn: 
  error: true
  failure: true
```

@jessesuen I can also implement this behavior for steps.

edlee2121 commented:
This looks good to me. @jessesuen?

schrodit commented:
@jessesuen @edlee2121 
I renamed the attribute to `continueOn` and implemented different behavior for NodePhase `error` and `fafiled`.

I also added the functionality to steps.

Can you please review?

schrodit commented on pkg/apis/workflow/v1alpha1/types.go:
```diff
@@ -940,3 +948,33 @@ func (wf *Workflow) NodeID(name string) string {
+// It can be specified if the workflow should continue when the pod errors, fails or both.
+type ContinueOn struct {
+	Error  *bool `json:"error"`
+	Failed *bool `json:"failed"`
```
done
